ed25519-donna-portable.h

```diff
@@ -35,9 +35,16 @@
 		typedef struct uint128_t {
 			uint64_t lo, hi;
 		} uint128_t;
-		#define mul64x64_128(out,a,b) out.lo = _umul128(a,b,&out.hi);
-		#define shr128_pair(out,hi,lo,shift) out = __shiftright128(lo, hi, shift);
-		#define shl128_pair(out,hi,lo,shift) out = __shiftleft128(lo, hi, shift);
+		#if defined(CPU_X86_64)
+			#define mul64x64_128(out,a,b) out.lo = _umul128(a,b,&out.hi);
+			#define shr128_pair(out,hi,lo,shift) out = __shiftright128(lo, hi, shift);
+			#define shl128_pair(out, hi, lo, shift) out = __shiftleft128(lo, hi, shift);
+		#else
+			#define mul64x64_128(out,a,b) out.lo = ((uint64_t)(a))*(b); out.hi = __umulh(a, b);
+			// note: these are special purpose functions where shift is expected to be < 64 and != 0
+			#define shr128_pair(out,hi,lo,shift) out = ((lo) >> shift) | ((hi) << (64 - shift));
+			#define shl128_pair(out,hi,lo,shift) out = (lo) << shift;
```

Comment:
@matbech  are you relay sure `shl128_pair` is correct?
Shouldn't it be implemented as 
```
#define shl128_pair(out,hi,lo,shift) out = ((hi) << shift) | ((lo) >> (64 - shift));
```
https://docs.microsoft.com/en-us/cpp/intrinsics/shiftleft128?view=vs-2019 states 
> Shifts a 128-bit quantity, represented as two 64-bit quantities LowPart and HighPart, to the left by a number of bits specified by Shift and returns the high 64 bits of the result.

Reply:
You are correct, both macros are incorrect.
#define shr128_pair(out,high,low,shift) out.lo = ((low) >> shift) | ((high) << (64 - shift)); out.hi =  ((high) >> shift);
#define shl128_pair(out,high,low,shift) out.lo = (low) << shift; out.hi = ((high) << shift) | ((low) >> (64 - shift));